File: TopCoder/python/SortedList.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next:ListNode = None


def merge(first_half: ListNode, second_half: ListNode) -> ListNode:

    dummy: ListNode = ListNode(-1)
    current: ListNode = dummy

    while first_half and second_half:
        if first_half.val < second_half.val:
            current.next = first_half
            first_half = first_half.next
            current = current.next
        else:
            current.next = second_half
            second_half = second_half.next
            current = current.next

    if first_half:
        current.next = first_half

    if second_half:
        current.next = second_half
    return dummy.next


class Solution:
    def sortList(self, head: ListNode) -> ListNode:
        if not head or not head.next:
            return head

        mid: ListNode = self.findMid(head)
        firstHalf : ListNode = self.sortList(head)
        secondHalf: ListNode = self.sortList(mid)

        return merge(firstHalf, secondHalf)

    def findMid(self, head) -> ListNode:
        slow: ListNode = head
        fast: ListNode = head

        while fast and fast.next:
            fast = fast.next.next
            if not fast or not fast.next:
                temp: ListNode = slow.next
                slow.next = None
                return temp
            else:
                slow = slow.next
        return slow


if __name__ == "__main__":
    head: ListNode = ListNode(4)
    head.next = ListNode(2)
    head.next.next = ListNode(1)
    head.next.next.next = ListNode(3)
    sol = Solution()
    result = sol.sortList(head)
    print(result)

File: TopCoder/src/main/java/com/leetcode/linkedlist/SortedList.java
package com.leetcode.linkedlist;

//https://leetcode.com/problems/sort-list/discuss/409256/Clean-and-organized-python-code
public class SortedList {

    public ListNode sortList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode mid = split(head);
        ListNode firstHalf = sortList(head);
        ListNode secondHalf = sortList(mid);
        return merge(firstHalf, secondHalf);
    }

    private ListNode merge(ListNode firstHalf, ListNode secondHalf) {
        ListNode dummy = new ListNode(-1);
        ListNode current = dummy;

        while (firstHalf != null && secondHalf != null) {
            if (firstHalf.val < secondHalf.val) {
               current.next = firstHalf;
               firstHalf = firstHalf.next;
               current = current.next;
            } else {
                current.next = secondHalf;
                secondHalf = secondHalf.next;
                current = current.next;
            }
            current.next = null;
        }

        if (firstHalf != null) {
            current.next = firstHalf;

        }

        if (secondHalf != null) {
            current.next = secondHalf;
        }
        return dummy.next;
    }

    private ListNode split(ListNode head) {
        ListNode slow = head;
        ListNode fast = slow;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            if (fast == null || fast.next == null) {
                ListNode temp = slow.next;
                slow.next = null;
                return temp;
            } else {
                slow = slow.next;
            }
        }
        return slow;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(4);
        head.next = new ListNode(2);
        head.next.next = new ListNode(1);
        head.next.next.next = new ListNode(3);

        ListNode sorted = new SortedList().sortList(head);

        while (sorted != null) {
            System.out.println(sorted.val);
            sorted = sorted.next;
        }
    }
}
